Use queues to process CAN frames in order

trnila opened the pull request:
CAN frames are currently put into stack/LIFO so last CAN frame is processed as a first. This behavior causes sending older data after newer one for example.

With queues we can process first CAN frame as a first, but then in depends on CAN driver and message ID priority.
Few other things were fixed/improved:
- incomplete CAN frame received over UDP is dropped (wrong DLC without data can lead to memory issues from CAN driver for example)
- CAN driver tx function (`cnl_can_tx_fn`) can return `false` so frame transmission will be retried next time instead of dropping it
- `flags` is zero if there are no flags

trnila committed:
use queue to send/receive frames in order

First received frame is processed as first

trnila committed:
dont drop frame if CAN driver cannot send it

CAN frame is kept in queue until it can be sent on the bus

trnila committed:
check if received UDP contains full header

trnila committed:
simplify parsing received UDP datagram by using index

trnila committed:
cannelloni: dont add incomplete frames received over UDP into queue

Incomplete frames can have wrong dlc leading to application crash.
The frame is at first validated and then new frame in the queue is
created.

mguentner commented on lwip/cannelloni.c:
```diff
@@ -57,112 +96,117 @@ void handle_cannelloni_frame(void *arg, struct udp_pcb *pcb, struct pbuf *p, con
+    memcpy(data, frame->data, canfd_len(frame));
+    data+=canfd_len(frame);
+    frameCount++;
+  } while (frame = queue_take(&handle->rx_queue));
```
Not introduced by you, but probably a good idea to fix when touching this part:

If this function is not called often enough, the resulting `dataPacket` might be larger than the MTU.
There should be another condition that leaves this queue, i.e. not sending the whole queue.
I suggest to add a simply return code indicating whether `transmit_udp_frame` should be called again to drain the queue.

trnila replied:
`transmit_udp_frame` is now called repeatedly when there are more frames in queue and `pbuf` allocation was successful.

Did the same thing also for the other queue - we are putting CAN frames to CAN driver (`can_tx_fn`) until it stops processing them (because mailboxes are full etc)

mguentner commented on lwip/cannelloni.c:
```diff
@@ -57,112 +96,117 @@ void handle_cannelloni_frame(void *arg, struct udp_pcb *pcb, struct pbuf *p, con
+  if (frame) {
+    if (handle->Init.can_tx_fn(handle, frame)) {
+      queue_take(&handle->tx_queue);
+    }
```
:+1:

mguentner reviewed (commented):
Thanks for the contribution. Just a small improvement wish from my side!

trnila committed:
transmit_udp_frame: dont access pbuf if allocation failed

trnila committed:
transmit_udp_frame: simplify building UDP datagram by using indexes instead of pointer arithmetics

trnila committed:
transmit_udp_frame: check if CAN frame fits into current UDP datagram to prevent memory overflow

First CAN frame is also not taken out of queue until we have allocated
UDP datagram to prevent frame loss.

trnila committed:
Drain RX queue by sending multiple UDP datagrams

transmit_udp_frame returns TRUE if there are more CAN frames and pbuf
allocation didn't fail

trnila committed:
Drain TX queue until CAN driver cannot process more frames

trnila committed:
Removed extra semicolon

mguentner commented:
Thanks, @trnila. I rebased the branch because I discovered that the original file were `CRLF` and I wanted to this PR to be already with just `LF`.
